Algorithm to calculate helium-3 analyser efficiency

jclarkeSTFC opened the pull request:
### Description of work

This is a new algorithm that will calculate the efficiency of a helium-3 analyser for a given run, which is required for the polarised SANS reduction. It will use the gas absorption parameter (cell length x pressure) calculated using either the algorithm in #36939 or some other method.

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Added algorithm
- Added tests
- Added doc page
- Added a release note
- Added validator for spin state strings

Fixes #36140. 

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

Takes a normalised group workspace with four periods representing a run, combines this with the cell path length multiplied by the gas pressure (``GasPressureTimesCellLength``) to calculate the efficiency of the analyser.

If the transmission of the wanted spin state is T_para, and the transmission of the unwanted spin state is T_anti, then the polarization of of an unpolarised incoming beam after the analyser cell, P_cell is given by 

![image](https://github.com/mantidproject/mantid/assets/139879523/1edf96cd-96fb-4f20-8afc-c03c00b29f27)

The efficiency of the analyser cell is given by

![image](https://github.com/mantidproject/mantid/assets/139879523/973dc202-4b81-4ef5-9dce-e04a8274ef72)

If our four periods are T_{00}, T_{01}, T_{10}, T_{11}, with the subscript denoting the spin configuration, then T_{anti} = T_{01} + T_{10} and T_{para} = T_{00} + T_{11}, and we can calculate P_cell from the above equation. The value of pd is given by the input ``GasPressureTimesCellLength ``, and lambda is the wavelength of each bin,  so we fit to our calculated e_cell to give us p_He and hence the theoretical efficiency curve.

To calculate the error, De_cell, we need the error on p_He, Dp_He, and the error of pd, Dpd, given by ``GasPressureTimesCellLengthError``. The covariance between pd and p_He is assumed to be zero. Then the error is given by

![image](https://github.com/mantidproject/mantid/assets/139879523/531085cb-a05f-4a48-a0a0-e1b2828c84ae)

If is n_b the number of histogram bins used in the fit, then define n:=n_b-3, since we are fitting three parameters. Then the factor t_crit follows a t distribution with n degrees of freedom, and probability density function f_t(x,n). For a standard 68.3% (1-sigma) error the factor t_crit is given by the solution to

![image](https://github.com/mantidproject/mantid/assets/139879523/66475373-212e-42b4-b8aa-4d9581f2496a)

As the number of histogram bins used in the fit increases, t_crit tends to 1.

The new spin state validator, `SpinStateValidator`, checks that the input string contains the specified number of pairs containing either 0 or 1 for each character. This analyser algorithm currently only works with four spin states in the group workspace input, so will only accept a string with four (unique) spin states.

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

To test an idealised situation you can run this:

``` python
wsPara = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1-0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1')
wsPara1 = CloneWorkspace(wsPara)
wsAnti = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1+0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1')
wsAnti1 = CloneWorkspace(wsAnti)

grp = GroupWorkspaces([wsPara,wsAnti,wsPara1,wsAnti1])
eff = HeliumAnalyserEfficiency(grp, SpinStates='11,01,00,10')
```

If you want to test on some real data then you can run the attached file, [analyser_test.txt](https://github.com/mantidproject/mantid/files/14981116/analyser_test.txt). Let me know if you'd like other suitable runs to test. That test script also outputs the same calculation performed using the original code from Diego so you can compare results.

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Add algorithm.

This algorithm with calculate the efficiency of a helium analyser,
given the transmission of the empty cell and the gas absorption
(from gas length x pressure).

jclarkeSTFC committed:
Add some tests.

jclarkeSTFC committed:
Remove redundant output

jclarkeSTFC committed:
Add several more tests.

jclarkeSTFC committed:
Change T to mean instead of sum

After discussions with the SANS scientists, T should be the mean of
T_para and T_anti.

jclarkeSTFC committed:
Add error calculation and some tests

Add the errors to the output transmission workspaces, where the covariance
matrix for T_E and pd is used with the error on pHe to give the full
resulting error. A correction factor is also required using an
appropriate t-distribution to get the correct confidence intervals.

jclarkeSTFC committed:
Add docs page

jclarkeSTFC committed:
Add a few more comments

jclarkeSTFC committed:
Add check for fit success.

Also add option so that the algorithm can be told to continue regardless
of whether the fit succeeded or not.

jclarkeSTFC committed:
Use matrix notation for error calc in docs

jclarkeSTFC committed:
Use const strings for parameter names

jclarkeSTFC committed:
Added release notes

jclarkeSTFC committed:
Add a couple of includes

jclarkeSTFC committed:
Fix compiler warnings

jclarkeSTFC committed:
Rename some parameters to fix tests

jclarkeSTFC committed:
Catch exception by reference

rbauststfc reviewed (commented):
I'm really sorry to only do a partial review at this stage, I didn't get as much done as I'd anticipated. I've not tested the functionality yet or finished working through the code, but here are some thoughts and queries so far. I will continue reviewing over the next couple of days.

Thanks very much for the detailed PR write-up, it was really helpful for understanding all the background in the code!

jclarkeSTFC committed:
Move files to PolarizationCorrections folder

jclarkeSTFC committed:
Add PolarizationCorrections to category

jclarkeSTFC committed:
Rename output workspace parameter

jclarkeSTFC committed:
Rename val to validator

jclarkeSTFC committed:
Remove unnecessary default constructor

jclarkeSTFC committed:
Add check that input workspace exists

jclarkeSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/PolarisedSANS/HeliumAnalyserEfficiency.h:
```diff
@@ -0,0 +1,51 @@
+  /// Algorithm's version for identification overriding a virtual method
+  int version() const override { return 1; }
+  /// Algorithm's category for identification overriding a virtual method
+  const std::string category() const override { return "SANS"; }
```
Done

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+namespace PropertyNames {
+static const std::string INPUT_WORKSPACE = "InputWorkspace";
+static const std::string OUTPUT_T_WORKSPACE = "OutputTransmissionWorkspace";
+static const std::string P_HE = "HeliumPolarisation";
```
Changed to `OutputWorkspace`

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+
+void HeliumAnalyserEfficiency::init() {
+  // Declare required input parameters for algorithm and do some validation here
+  auto val = std::make_shared<CompositeValidator>();
```
Done

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+} // namespace PropertyNames
+
+/// Empty default constructor algorithm() calls the constructor in the base class
+HeliumAnalyserEfficiency::HeliumAnalyserEfficiency() : Algorithm() {}
```
I've deleted this constructor, seems to work no problem.

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+ */
+std::map<std::string, std::string> HeliumAnalyserEfficiency::validateInputs() {
+  std::map<std::string, std::string> errorList;
+  const auto ws = AnalysisDataService::Instance().retrieve(getProperty(PropertyNames::INPUT_WORKSPACE));
```
Added a check that the workspace actually exists.

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+ * of issues in the input properties.
+ */
+void HeliumAnalyserEfficiency::validateGroupInput() {
+  const auto results = validateInputs();
```
It looks like if you call the algorithm from a script, then it won't call `validateInputs`, but it will if you use the dialog window.

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+  val->add<HistogramValidator>();
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::INPUT_WORKSPACE, "", Direction::Input, val));
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_T_WORKSPACE, "T", Direction::Output));
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::P_HE, "p_He", Direction::Output));
```
I'm not sure yet which are going to be the most used outputs, because e.g. in the polarizer efficiency algorithm we're going to be using the `T_para` and `T_anti` workspaces.

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+using namespace API;
+
+namespace SpinConfigurations {
+static const std::string UpUp = "11";
```
I'll do this when I refactor that stuff into a separate place.

jclarkeSTFC commented on Framework/Algorithms/src/PolarisedSANS/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,356 @@
+
+  std::vector<std::string> initialSpinConfig{
+      {SpinConfigurations::UpUp, SpinConfigurations::UpDown, SpinConfigurations::DownUp, SpinConfigurations::DownDown}};
+  std::sort(initialSpinConfig.begin(), initialSpinConfig.end());
```
From the docs for `std::next_permutation` it looks like it returns permutations in lexicographical order, so if the initial vector isn't sorted then `next_permutation` will stop early, without giving all the possible permutations. But I'm thinking that it might be better to just use a custom validator that lets the user type in a string, instead of the combo box setup I used here. Although I'll probably try and manoeuvre in a use of this permutation stuff, seems like a waste otherwise.

jclarkeSTFC committed:
Tidy up main body of algorithm

Split the actual calculation into several separate methods, hopefully
it's now a bit clearer.

jclarkeSTFC committed:
Add new validator for spin states

Specifying and validating spin states is something we're going to need
to do in multiple places, so I've made a validator class to handle the
input, and also to help extracting the relevant workspaces. Should
avoid some code duplication.

jclarkeSTFC committed:
Change output to efficiency curve

Previously the output was the transmission curve, but output the
efficiency curve by default. The transmission curves are now optional
outputs. Calculating the efficiency curve within this algorithm means
I had to add the error calculation, because the error directly from
the fit won't include the error in pd.

rbauststfc commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,280 @@
+    heliumAnalyserEfficiency->setProperty("InputWorkspace", wsGrp->getName());
+    heliumAnalyserEfficiency->setProperty("Covariance", covariance);
+    heliumAnalyserEfficiency->execute();
+    TS_ASSERT_EQUALS(true, heliumAnalyserEfficiency->isExecuted());
```
Is there anything we can add here that would test that the covariance matrix has been used correctly by the algorithm (i.e. any data values that could be checked)? I think at the moment if a regression was introduced into how the algorithm uses the covariance matrix then this test would still pass.

jclarkeSTFC replied:
I've added a couple of tests that check the error output with different covariance matrices.

rbauststfc commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,280 @@
+      AnalysisDataService::Instance().remove(members[i]);
+    }
+
+    AnalysisDataService::Instance().remove(heliumAnalyserEfficiency->getProperty("OutputWorkspace"));
```
It's great to have `AnalysisDataService::Instance().clear()` in the test teardown method, so are all of these individual removes needed? Also, is there a way we could check some of the contents of the output workspaces to verify that they have been created successfully and pick up any regressions in future?

jclarkeSTFC replied:
I changed this test to just check for the workspaces existing in the ADS, instead of removing them as it goes, since they're going to get cleared out in the tear down anyway.

thomashampson commented:
pre-commit.ci run

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+  validator->add<HistogramValidator>();
+  declareProperty(
+      std::make_unique<WorkspaceProperty<>>(PropertyNames::INPUT_WORKSPACE, "", Direction::Input, validator));
+  declareProperty(std::make_unique<WorkspaceProperty<>>(PropertyNames::OUTPUT_WORKSPACE, "PA", Direction::Output));
```
I don't think we would normally provide a default for the output workspace, so I'd recommend that we just leave the default as an empty string here too. I found it a bit confusing when testing, so it would be good to be consistent with other algorithms.

jclarkeSTFC replied:
Done

rbauststfc reviewed (commented):
I've done some testing with the algorithm and it completes successfully, with appropriate validation when there are issues. Code changes are looking really good, I've added some comments, but none relate to functional issues.

One thing I noticed when testing is that there seems to the same issue as I've commented on on the transmission PR: https://github.com/mantidproject/mantid/pull/36939#pullrequestreview-1962704037. This might be something for both you and Caila to investigate together to see if we can understand what's behind this.

jclarkeSTFC committed:
Add empty string test cases

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/SpinStateValidatorTest.h:
```diff
@@ -0,0 +1,63 @@
+
+class SpinStateValidatorTest : public CxxTest::TestSuite {
+public:
+  void testSingleCorrectInputs() {
```
Done

jclarkeSTFC committed:
Rename pHe parameter

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+namespace PropertyNames {
+static const std::string INPUT_WORKSPACE = "InputWorkspace";
+static const std::string OUTPUT_WORKSPACE = "OutputWorkspace";
+static const std::string P_HE = "HeliumPolarisation";
```
I've renamed it to `HeliumAtomsPolarization`, which is what Krycka calls it, (with a z)

jclarkeSTFC committed:
Add descriptions to properties

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+  auto validator = std::make_shared<CompositeValidator>();
+  validator->add<WorkspaceUnitValidator>("Wavelength");
+  validator->add<HistogramValidator>();
+  declareProperty(
```
Done

jclarkeSTFC committed:
Move constant to top of file

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+  return true;
+}
+
+const double HeliumAnalyserEfficiency::ABSORPTION_CROSS_SECTION_CONSTANT = 0.0733;
```
Done

jclarkeSTFC committed:
Remove unnecessary method

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+bool HeliumAnalyserEfficiency::propertyHasValue(const std::string &property) {
+  std::string p = getPropertyValue(property);
+  if (p.empty()) {
+    setPropertyValue(property, "");
```
Done. When I first made the outputs optional I was getting an exception if I didn't set the output to something, but it seems fine now, not sure what happened there.

jclarkeSTFC committed:
Move optional output property code

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,409 @@
+  auto pCalcWorkspace = createEfficiencyWorkspace(pxd, pHe, pHeError, tCrit, wavelengthValues, pCalc);
+  setProperty(PropertyNames::OUTPUT_WORKSPACE, pCalcWorkspace);
+
+  if (propertyHasValue(PropertyNames::P_HE)) {
```
Done

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,280 @@
+    covariance->addColumn("double", "b");
+    covariance->appendRow();
+
+    // Test covariance matrix with incorrect size
```
Done

jclarkeSTFC committed:
Split up covariance matrix test

jclarkeSTFC committed:
Delete default output workspace

jclarkeSTFC committed:
Split helper method into separate file

Move the WorkspaceForSpinState method into a separate helper file, and
added some tests.

jclarkeSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/PolarizationCorrections/SpinStateValidator.h:
```diff
@@ -0,0 +1,40 @@
+public:
+  SpinStateValidator(std::unordered_set<int> allowedNumbersOfSpins);
+  Kernel::IValidator_sptr clone() const override;
+  static API::MatrixWorkspace_sptr WorkspaceForSpinState(API::WorkspaceGroup_sptr group, std::string spinStateOrder,
```
I split that method off into a separate file in the same folder

jclarkeSTFC committed:
Rename stop on fit error option

jclarkeSTFC committed:
Add a couple of covariance tests

Change test example generation method to generate a realistic set of
numbers, then add a couple of tests that check the covariance matrix
does something. We need realistic numbers because if the pHe fit is
zero then the covariance will have no effect on the resulting errors.

jclarkeSTFC committed:
Change test to leave ADS removal to tear down

The tear down is going to remove all the workspaces anyway, so in this
test just check if the workspaces exist.

jclarkeSTFC committed:
Change output from polarization to efficiency

We want the efficiencies for the following algorithm, so change the main
output (but keep the polarization as an optional output). Updated the
docs, added a few tests, updated the algorithm.

jclarkeSTFC committed:
FIx typo in doc test output

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,493 @@
+  fit->initialize();
+  fit->setProperty("Function", "name=UserFunction,Formula=tanh(" + std::to_string(mu) + "*phe*x),phe=0.1");
+  fit->setProperty("InputWorkspace", p);
+  const double startLambda = getProperty("StartLambda");
```
It would be good to use the property name constants for start and end lambda here, if we can.

jclarkeSTFC replied:
Done

rbauststfc reviewed (commented):
This is increasingly looking really good. I'm afraid I do have a few final comments, sorry. Most are focussed on a few further changes to the tests to help with maintainability. There are a couple of minor suggestions in the main algorithm - feel free to go with your own preference on whether to address these. There were also a couple of potential bugs I noticed, one to do with the ignore fit error property (I haven't actually been able to test for sure because I couldn't find settings that resulted in a poor fit, so let me know if there's a good way to check this) and one when finding the workspace in the group that matches the spin configuration.

Just let me know if you want to chat anything through at all.

jclarkeSTFC committed:
No need for intermediate step here

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,493 @@
+  auto groupWorkspace = createChildAlgorithm("GroupWorkspaces");
+  groupWorkspace->initialize();
+  std::vector<std::string> wsToGroupNames(4);
+  std::transform(wsVector.cbegin(), wsVector.cend(), wsToGroupNames.begin(),
```
Yeah there's no need for the intermediate step, I've changed it to just get the name immediately.

jclarkeSTFC committed:
Fix logic following refactor

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,493 @@
+  fit->setProperty("CreateOutput", true);
+  fit->execute();
+
+  const bool stopOnFitError = getProperty(PropertyNames::IGNORE_FIT_QUALITY_ERROR);
```
I've fixed the logic, this was a hangover from changing the input argument to match the other algorithm.

jclarkeSTFC committed:
Reduce scope of test helpers

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,414 @@
+    }
+  }
+
+  WorkspaceGroup_sptr createExampleGroupWorkspace(const std::string &name, const std::string &xUnit = "Wavelength",
```
Done

jclarkeSTFC committed:
Remove redundant test case

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,414 @@
+    TS_ASSERT_THROWS(heliumAnalyserEfficiency->setProperty("SpinStates", "10,01"), std::invalid_argument &);
+    TS_ASSERT_THROWS(heliumAnalyserEfficiency->setProperty("SpinStates", "00,00,11,11"), std::invalid_argument &);
+    TS_ASSERT_THROWS(heliumAnalyserEfficiency->setProperty("SpinStates", "02,20,22,00"), std::invalid_argument &);
+    TS_ASSERT_THROWS_NOTHING(heliumAnalyserEfficiency->setProperty("SpinStates", "00,11,01,10"));
```
Done

jclarkeSTFC committed:
Reduce scope of test helper method

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/SpinStateValidatorTest.h:
```diff
@@ -0,0 +1,63 @@
+    checkAllInputs(validator, correctInputs, true);
+  }
+
+  void checkAllInputs(const std::shared_ptr<SpinStateValidator> validator, const std::vector<std::string> &inputsToTest,
```
Donee

jclarkeSTFC committed:
Use property name constants

jclarkeSTFC committed:
Add method for splitting spin states

Avoid problems with leading/trailing whitespace in the spin state string
by using a single utility method to split the string up into individual
spin states.

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/PolarizationCorrectionsHelpers.cpp:
```diff
@@ -0,0 +1,36 @@
+size_t IndexOfWorkspaceForSpinState(API::WorkspaceGroup_sptr group, const std::string &spinStateOrder,
+                                    const std::string &targetSpinState) {
+  std::vector<std::string> spinStateVector;
+  boost::split(spinStateVector, spinStateOrder, boost::is_any_of(","));
```
I've added a utility method to split the spin state string, should avoid this problem. There's an extra test as well.

jclarkeSTFC committed:
Add property groups

jclarkeSTFC committed:
Refactor tests

jclarkeSTFC commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -0,0 +1,414 @@
+    std::vector<double> x{1, 2, 3, 4, 5};
+    std::vector<double> y{1, 4, 9, 16, 25};
+
+    MatrixWorkspace_sptr ws1 = generateWorkspace("ws1", x, y);
```
Done

jclarkeSTFC committed:
Fix cppcheck warning

jclarkeSTFC committed:
Loosen tolerance to prevent intermittency

jclarkeSTFC committed:
Change way algorithm works

Calculate the efficiency directly instead of first calculating the
polarization. Remove all optional outputs. Update docs.

jclarkeSTFC committed:
Use same bins for output

Now the algorithm will output efficiency for the same bins as the
input workspace(s), not the range defined by the fit. I also added a
test to check for this behaviour.

cailafinn reviewed (dismissed):
This looks good to me now. Nice work!

thomashampson commented on Framework/Algorithms/src/PolarizationCorrections/SpinStateValidator.cpp:
```diff
@@ -0,0 +1,57 @@
+
+std::string SpinStateValidator::checkValidity(const std::string &input) const {
+  if (input.empty())
+    return "Enter a spin state string, it should be a comma-separated list of spin states, e.g. 01, 01";
```
is the example given here an invalid one?
```suggestion
    return "Enter a spin state string, it should be a comma-separated list of spin states, e.g. 01, 11";
```

jclarkeSTFC committed:
Give correct example in validator text

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

thomashampson commented on Framework/Algorithms/inc/MantidAlgorithms/PolarizationCorrections/PolarizationCorrectionsHelpers.h:
```diff
@@ -0,0 +1,24 @@
+
+namespace Mantid::Algorithms {
+namespace PolarizationCorrectionsHelpers {
+MANTID_ALGORITHMS_DLL API::MatrixWorkspace_sptr WorkspaceForSpinState(API::WorkspaceGroup_sptr group,
```
I think the functions in here should be camelCase

jclarkeSTFC committed:
We now only have two parameters instead of three

In the old formulation we used T_E, pd, and p_He, but now we only have
pd and p_He. Hence the t distribution needs to be adjusted to only
lose 2 degrees of freedom instead of 3.

jclarkeSTFC committed:
Change method to lower camel case

cailafinn reviewed (changes requested):
Works well, just had a few more code comments.

jclarkeSTFC committed:
Use constexpr instead of const

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

jclarkeSTFC committed:
Tidy up some code

Removed a couple of unused arguments, removed an unnecessary entry to
the log (it was throwing an exception anyway).

jclarkeSTFC commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -0,0 +1,284 @@
+  const std::string &status = fit->getProperty("OutputStatus");
+  if (!ignoreFitQualityError && (!fit->isExecuted() || status != "success")) {
+    auto const &errMsg{"Failed to fit to data in the calculation of p_He: " + status};
+    g_log.error(errMsg);
```
I've removed the call to the logger here. When running from the dialog it does end up here https://github.com/mantidproject/mantid/blob/3b18faf9e8b3fea73910fa8f356e545bc65bed43/Framework/API/src/Algorithm.cpp#L1263 when throwing an exception so not sure why it's not logging in that case. I had a look at a few other algorithms with overrides to `processGroups` and I couldn't see anything special for most of them, maybe this is a problem for them as well.

cailafinn reviewed (approved):
This worked well and looks good to me :shipit:
